bibliopixel/drivers/spi_driver_base.py

```diff
@@ -1,75 +1,158 @@
-        super().__init__(num, c_order=c_order, gamma=gamma)
+class SpiBaseInterface(object):
+    """ abstract class for different spi backends"""
 
-        self.dev = dev
-        self.use_py_spi = use_py_spi
-        self._spiSpeed = SPISpeed
-        self._open = open
+    def __init__(self, dev, SPISpeed):
+        self._dev = dev
+        self._spi_speed = SPISpeed
 
-        # File based SPI requires a bytearray so we have to overwrite _buf
-        if self.use_py_spi:
-            a, b = -1, -1
-            d = self.dev.replace('/dev/spidev', '')
-            s = d.split('.')
-            if len(s) == 2:
-                a = int(s[0])
-                b = int(s[1])
+    def send_packet(self, data):
+        raise NotImplementedError
 
-            if a < 0 or b < 0:
-                error(BAD_FORMAT_ERROR)
+    def compute_packet(self, data):
+        return data
 
-            self._bootstrapSPIDev()
-            import spidev
-            self.spi = spidev.SpiDev()
-            self.spi.open(a, b)
-            self.spi.max_speed_hz = int(self._spiSpeed * 1000000.0)
-            log.info('py-spidev speed @ %.1f MHz',
-                     (float(self.spi.max_speed_hz) / 1000000.0))
-        else:
-            self._buf = bytearray(self._buf)
-            self.spi = self._open(self.dev, 'wb')
-
-    def _bootstrapSPIDev(self):
-        if self.use_py_spi:
-            try:
-                import spidev
-            except:
-                error(CANT_IMPORT_SPIDEV_ERROR)
-
-        if not os.path.exists(self.dev):
+
+# partial source @see https://armbedded.taskit.de/node/318
+# @see <linux/spi/spidev.h>
+# hex numbers from good old strace(1)
+SPI_IOC_WR_MODE = 0x40016b01
+SPI_IOC_RD_MODE = 0x80016b01
+SPI_IOC_WR_BITS_PER_WORD = 0x40016b03
+SPI_IOC_RD_BITS_PER_WORD = 0x80016b03
+SPI_IOC_WR_MAX_SPEED_HZ = 0x40046b04
+SPI_IOC_RD_MAX_SPEED_HZ = 0x80046b04
+
+
+class SpiFileInterface(SpiBaseInterface):
+    """ using os open/write to send data"""
+
+    def __init__(self, **kwargs):
+        super().__init__(**kwargs)
+        if not os.path.exists(self._dev):
             error(CANT_FIND_ERROR)
 
+        self._fnctl = fcntl
+        self._spi = open(self._dev, "wb")
+
+        speed = self._set_speed(self._spi_speed)
+
+        bits = self._set_bits_per_word(8)
+        log.info('file io spi speed @ %.1f MHz, %d bits per word', speed / 1e6, bits)
+
+    def _set_speed(self, speed):
+        speed_b = struct.pack(">I", int(speed * 1e9))  # unint32
+        self._fnctl.ioctl(self._spi, SPI_IOC_WR_MAX_SPEED_HZ, speed_b)
+
+        self._fnctl.ioctl(self._spi, SPI_IOC_RD_MAX_SPEED_HZ, speed_b)
+        speed = struct.unpack(">I", speed_b)[0] / 1000
+        return speed
+
+    def _set_bits_per_word(self, bits):
+        bits_b = struct.pack("B", bits)  # uint8
+        self._fnctl.ioctl(self._spi, SPI_IOC_WR_BITS_PER_WORD, bits_b)
+
+        self._fnctl.ioctl(self._spi, SPI_IOC_RD_MAX_SPEED_HZ, bits_b)
```

Comment:
[This extraction of common code](https://github.com/rec/BiblioPixel/commit/c61a20261bcef79d268dc135e2f0c0e92da84bc4) would make that sort of error impossible and make the code somewhat shorter too.